AI/VCAI/VCAI.cpp

```diff
@@ -122,11 +123,19 @@ void VCAI::heroMoved(const TryMoveHero & details)
 		const CGObjectInstance *o1 = frontOrNull(cb->getVisitableObjs(from)),
 			*o2 = frontOrNull(cb->getVisitableObjs(to));
 
-		if(o1 && o2 && o1->ID == Obj::SUBTERRANEAN_GATE && o2->ID == Obj::SUBTERRANEAN_GATE)
+		auto t1 = dynamic_cast<const CGTeleport *>(o1);
+		auto t2 = dynamic_cast<const CGTeleport *>(o2);
+		if(t1 && t2)
 		{
-			knownSubterraneanGates[o1] = o2;
-			knownSubterraneanGates[o2] = o1;
-            logAi->debugStream() << boost::format("Found a pair of subterranean gates between %s and %s!") % from % to;
+			if(ETeleportChannelType::BIDIRECTIONAL == cb->getTeleportChannelType(t1->channel))
+			{
+				if(o1->ID == Obj::SUBTERRANEAN_GATE)
+				{

```

Comment:
Here should be something about that o2 is also SUBTERRANEAN_GATE, assert, additional check, etc and a comment _why_ is must be so.